lib/config/dynamic/admiral/source.go

```diff
@@ -0,0 +1,396 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package admiral
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/go-openapi/runtime"
+	rtclient "github.com/go-openapi/runtime/client"
+	strfmt "github.com/go-openapi/strfmt"
+	"github.com/mitchellh/mapstructure"
+	"github.com/pkg/errors"
+
+	"github.com/vmware/govmomi/vim25/types"
+	vchcfg "github.com/vmware/vic/lib/config"
+	"github.com/vmware/vic/lib/config/dynamic"
+	"github.com/vmware/vic/lib/config/dynamic/admiral/client"
+	"github.com/vmware/vic/lib/config/dynamic/admiral/client/config_registries"
+	"github.com/vmware/vic/lib/config/dynamic/admiral/client/projects"
+	"github.com/vmware/vic/lib/config/dynamic/admiral/client/resources_compute"
+	"github.com/vmware/vic/lib/config/dynamic/admiral/models"
+	"github.com/vmware/vic/pkg/vsphere/session"
+	"github.com/vmware/vic/pkg/vsphere/tags"
+	"github.com/vmware/vic/pkg/vsphere/vm"
+)
+
+const (
+	VicProductCategory = "VsphereIntegratedContainers"
+	ProductVMTag       = "ProductVM"
+	admiralEndpointKey = "guestinfo.vicova.admiral.endpoint"
+
+	clusterFilter = "(address eq '%s' and customProperties.__containerHostType eq 'VCH')"
+)
+
+// #nosec
+const admiralTokenKey = "guestinfo.vicova.engine.token"
+
+var (
+	trueStr        = "true"
+	projectsFilter = "customProperties.__enableContentTrust eq 'true'"
+)
+
+// NewSource creates a new Admiral dynamic config source. sess
+// is a valid vsphere session object. vchID is a unique identifier
+// that will be used to lookup the VCH in the Admiral instance; currently
+// this is a URI to the docker endpoint in the VCH.
+func NewSource(sess *session.Session, vchID string) dynamic.Source {
+	return &source{
+		d:     &productDiscovery{},
+		sess:  sess,
+		vchID: vchID,
+	}
+}
+
+type source struct {
+	mu      sync.Mutex
+	d       discovery
+	sess    *session.Session
+	vchID   string
+	c       *client.Admiral
+	lastCfg *vchcfg.VirtualContainerHostConfigSpec
+}
+
+// Get returns the dynamic config portion from an Admiral instance. For now,
+// this is empty pending details from the Admiral team.
+func (a *source) Get(ctx context.Context) (*vchcfg.VirtualContainerHostConfigSpec, error) {
+	var err error
+	if err = a.discover(ctx); err != nil {
+		return nil, err
+	}
+
+	var projs []string
+	projs, err = a.projects(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var wl []string
+	wl, err = a.whitelist(ctx, projs)
+	if err != nil {
+		return nil, err
+	}
+
+	newCfg := &vchcfg.VirtualContainerHostConfigSpec{
+		Registry: vchcfg.Registry{RegistryWhitelist: wl},
+	}
+
+	err = a.diff(newCfg)
+	a.lastCfg = newCfg
+	return newCfg, err
+}
+
+func (a *source) diff(newCfg *vchcfg.VirtualContainerHostConfigSpec) error {
+	if a.lastCfg == nil {
+		return nil
+	}
+
+	if newCfg == nil {
+		return dynamic.ErrConfigNotModified
```

Comment:
Yes; currently we don't hit that case since `newCfg` is never `nil`. But I'll fix it.